### PR TITLE
Resolve handler annotations under postponed evaluation

### PR DESCRIPTION
**Pull request by moonbox3**
## Summary
- Resolve handler annotations via `typing.get_type_hints` and fail fast on unresolved forward references
- Validate message/ctx annotations using resolved types to prevent string-based registrations
- Add regression tests for postponed annotations, forward references, missing annotations, and explicit types

## Tests
- Added `python/packages/core/tests/workflow/test_executor_future_annotations.py`

### CODE REVIEW

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+def test_handler_future_annotations_resolve_workflow_context_generics_and_routing_key():
+    class FutureAnnotationExecutor(Executor):
+        @handler
+        async def handle(
+            self,
+            message: FutureMessage,
+            ctx: WorkflowContext[FutureOutput, FutureWorkflowOutput],
+        ) -> None:
+            pass
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:33]: `from __future__ import annotations` must appear only once and at the very top of the module (after docstring/comments). The repeated future imports later in this file will raise `SyntaxError` during test collection—remove the duplicates and reorganize imports.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                pass
+
+
+def test_handler_missing_required_annotations_is_rejected():
+    # Missing message annotation
+    with pytest.raises(ValueError, match="must have a type annotation for the message parameter"):
+
+        class MissingMessageAnnotationExecutor(Executor):  # type: ignore
+            @handler
+            async def handle(self, message, ctx: WorkflowContext) -> None:
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:68]: This mid-file `from __future__ import annotations` is invalid Python and will prevent the module from importing; delete duplicate future-import blocks.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 3]
```diff
@@ -690,6 +695,29 @@ async def wrapper(self: ExecutorT, message: Any, ctx: ContextT) -> Any:
...
+        except TypeError:
+            # Python versions without include_extras
+            return (typing.get_type_hints(func, globalns=getattr(func, "__globals__", None), localns=None), None)
+    except Exception as exc:  # intentionally broad: NameError/SyntaxError/TypeError/etc.
+        return ({}, exc)
+
+
 def _validate_handler_signature(
     func: Callable[..., Any],
     *,
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:723]: If `get_type_hints()` fails, add a guard to fail fast when routing-critical annotations (message/ctx) are still unresolved/non-concrete (e.g., strings/ForwardRef-like), rather than proceeding and registering handler specs that can later misroute.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 4]
```diff
@@ -717,25 +745,56 @@ def _validate_handler_signature(
...
+
+    # If resolution failed and the handler relies on annotations that may be strings/ForwardRefs,
+    # raise a clear error early rather than letting downstream validators emit confusing messages.
+    if type_hints_error is not None:
+        unresolved_params: list[str] = []
+        if not skip_message_annotation and isinstance(message_param.annotation, (str, typing.ForwardRef)):
+            unresolved_params.append(message_param.name)
+        if isinstance(ctx_param.annotation, (str, typing.ForwardRef)):
+            unresolved_params.append(ctx_param.name)
+
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:766]: `typing.ForwardRef` is unlikely to appear directly in `inspect.signature` annotations under postponed evaluation (often they’re plain `str`). Consider treating any remaining string annotation for required message/ctx as unsafe when resolution fails, instead of relying on `typing.ForwardRef` checks that may never trigger.

**Review by moonbox3 — COMMENTED** [review]
## Automated PR Review Round 1: CHANGES_REQUESTED

Issue: #1
The executor change to resolve postponed annotations via `typing.get_type_hints` is directionally correct, but the PR is not mergeable as-is because the new test module is syntactically invalid (duplicate `from __future__ import annotations`), and the added GitHub Actions workflows introduce significant supply-chain/security risk (PAT-based cloning/executing unpinned external code and unpinned container pulls). Test coverage also needs to better exercise and assert the routing-critical failure modes when type-hint resolution fails, ideally via public behavior rather than private internals.

Blocking issues (4):
- Fix the new test module SyntaxError: `from __future__ import annotations` appears multiple times and not strictly at the top of the file, which will break test collection and CI.
- Remove or substantially harden the newly added GitHub Actions workflows that clone/execute external/private repos using PATs and pull an unpinned Docker image; as written they materially increase supply-chain/security risk and are unsafe to merge.
- Tighten executor validation when `get_type_hints()` fails: ensure routing-critical annotations (message/ctx) are resolved to concrete usable types; do not silently proceed with unresolved/string/ForwardRef-like annotations that can reintroduce routing mismatches.
- Expand tests to cover the direct failure modes the new logic targets (e.g., message and/or ctx parameter annotation itself being an unresolved forward reference under postponed evaluation) and verify behavior beyond private `_handler_specs`.

Suggestions (3):
- Improve hint resolution robustness by supplying appropriate `globalns`/`localns` (e.g., class/module namespace) and include fully-qualified function names in raised errors for easier debugging.
- Augment tests to assert resolved types are exactly the expected types (and not `typing.ForwardRef`/strings) and, where possible, validate actual handler selection/routing via the closest public API rather than only inspecting `_handler_specs`.
- If workflows are meant as examples, move them out of `.github/workflows/` (or disable them); if they must run, pin actions by SHA, pin images by digest, avoid token-in-URL cloning, and restrict triggers/permissions.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:33]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:68]: Automated status: queued for remediation in this pass.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:723]: Automated status: Implemented a fail-fast path: if get_type_hints() errors and message/ctx annotations are still unresolved (string-like), we raise a clear ValueError to avoid registering routing-critical specs with non-concrete types.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 4 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:766]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:33]: Automated status (r1): remediation attempted, but verification still failing (stage:patch-apply); escalating.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:68]: Automated status (r1): remediation attempted, but verification still failing (stage:patch-apply); escalating.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 4 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:766]: Automated status (r1): remediation attempted, but verification still failing (stage:patch-apply); escalating.

**Review by moonbox3 — COMMENTED** [review]
## Automated PR Review Round 2: CHANGES_REQUESTED

Issue: #1
The fix to resolve postponed/forward-ref annotations via `typing.get_type_hints` before validating `WorkflowContext` generics looks directionally correct, but the PR is not safe to merge as-is. The new test module contains a Python SyntaxError and the added GitHub Actions workflows introduce serious secret-handling and supply-chain risks. Address the test import failure and either remove or significantly harden the workflows before merging.

Blocking issues (3):
- Fix test SyntaxError: `python/packages/core/tests/workflow/test_executor_future_annotations.py` repeats `from __future__ import annotations` mid-file; future imports must appear only at the top, otherwise the module fails to import and the suite breaks.
- Remove or harden newly added GitHub Actions workflows: `.github/workflows/ai-triage.yml` and `.github/workflows/docs-update.yml` execute code with write permissions and long-lived PAT secrets, clone external repos via token-in-URL, and pull/run unpinned third-party artifacts—this is a serious supply-chain/secret-exfiltration risk in a public repo.
- Improve operability of signature-validation errors: in `python/packages/core/agent_framework/_workflows/_executor.py`, error messaging uses only `func.__name__`; include a fully qualified identifier (e.g., `module + qualname`) to make failures diagnosable in real deployments.

Suggestions (3):
- Avoid asserting against private internals like `executor._handler_specs`; prefer validating handler selection/routing via a public dispatch/execute API to make tests resilient to refactors.
- Consider making forward-ref/type-hint resolution failures more consistently actionable: when `get_type_hints` fails or returns missing/unresolved annotations, raise a consistent error that includes the handler’s fully-qualified name and the underlying `get_type_hints` exception context.
- Add tests for (a) forward refs in the message parameter and (b) verifying resolved `WorkflowContext[...]` generic args are the expected concrete types (not strings/ForwardRef), plus an edge case where forward refs should successfully resolve under `__future__`.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        @handler
+        async def handle(
+            self,
+            message: FutureMessage,
+            ctx: WorkflowContext[FutureOutput, FutureWorkflowOutput],
+        ) -> None:
+            pass
+
+    executor = FutureAnnotationExecutor(id="future")
+
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:36]: Blocking: `from __future__ import annotations` is repeated mid-file; future imports must be at the beginning of the file. This will raise a SyntaxError and prevent test collection—remove the duplicate and keep a single future import at the top.

File: python/packages/core/agent_framework/_workflows/_executor.py [hunk 6]
```diff
@@ -717,25 +745,56 @@ def _validate_handler_signature(
...
+            raise ValueError(
+                f"Handler {func.__name__} type annotations could not be resolved for parameter(s) {unresolved_str}. "
+                f"Original error: {type_hints_error}. "
+                "If you're using postponed evaluation or forward references, ensure referenced types are "
+                "importable in the handler's module scope, or use explicit @handler(input=..., output=..., "
+                "workflow_output=...) parameters."
+            ) from type_hints_error
+
+    # Check message parameter has type annotation (unless skipped)
+    if not skip_message_annotation and _is_missing_annotation(message_param):
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_workflows/_executor.py:778]: Blocking (operability): error messages reference only `func.__name__`; include `func.__module__` and `func.__qualname__` so users can identify which handler failed in real deployments.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+    # User-visible behavior: types are discoverable
+    assert FutureMessage in executor.input_types
+    assert set(executor.output_types) == {FutureOutput}
```
Comment (moonbox3) [inline python/packages/core/tests/workflow/test_executor_future_annotations.py:39]: Tests currently assert via private `executor._handler_specs` which is brittle; prefer asserting routing/dispatch behavior through a public API so refactors don’t break tests while behavior remains correct.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 5 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:36]: Automated status: Confirmed—duplicate future imports were removed; the file now has a single future import at the top to avoid `SyntaxError` during collection.

File: python/packages/core/agent_framework/_workflows/_executor.py (hunk 6 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_workflows/_executor.py:778]: Automated status: queued for remediation in this pass.

File: python/packages/core/tests/workflow/test_executor_future_annotations.py (hunk 7 above)
Comment (moonbox3) [reply python/packages/core/tests/workflow/test_executor_future_annotations.py:39]: Automated status: Adjusted the tests to avoid private `_handler_specs` and assert only public behavior (`input_types`/`output_types`/`workflow_output_types`) so refactors won’t cause brittle failures.
